math/miller_rabin.cpp

```diff
@@ -0,0 +1,185 @@
+/**
+ * Copyright 2020 @author tjgurwara99
+ * @file
+ *
+ * A basic implementation of Miller-Rabin primality test.
+ */
+
+#include <iostream>
+#include <random>
+#include <vector>
+#include <cassert>
+
+/**
+ * Function to give a binary representation of a number in reverse order
+ * @param integer number that we want to convert
+ * @return vector of the number input in reverse binary
+ */
+std::vector<int> reverse_binary(int num) {
+    std::vector<int> result;
+    int temp = num;
+    while (temp > 0) {
+        result.push_back(temp % 2);
+        temp = temp / 2;
+    }
+    return result;
+}
+
+/**
+ * Function for modular exponentiation.
+ * This function is an efficient modular exponentation function.
+ * It can be used with any big integer library such as Boost multiprecision
+ * to give result any modular exponentation problem reletively quickly.
+ * @param base that is being raised to a power as integer
+ * @param reverse binary of the power the base is being raised to
+ * @param modulo
+ * @return the modular exponentiation of a^(reverse_binary) % n
+ */
+int modular_exponentiation(int a, std::vector<int> rev_binary, int n) {
+    if (n == 1)
+        return 0;
+    int b = 1;
+    if (rev_binary.size() == 0)
+        return b;
+    int A = a;
+    if (rev_binary[0] == 1)
+        b = a;
+
+    for (std::vector<int>::iterator it = rev_binary.begin() + 1;
+         it != rev_binary.end(); ++it) {
+        A = A * A % n;
+        if (*it == 1)
+            b = A * b % n;
+    }
+    return b;
+}
+
+/**
+ * Function to raise a number to a power
+ * @param base
+ * @param power
+ * @return base^power
+ */
+int power(int num, int pow) {
```

Comment:
can this be replaced by `std::pow` from `<cmath>`?

Reply:
Yeah it can be but I thought that would create an unnecessary overhead. I can add it if that's preferred.

Reply:
in fact it is the opposite. please replace the custom function with the one from STL C++

Reply:
Oh ok, thanks!

Reply:
Apologies, I had forgotten that I changed my implementation that didn't require me to use this function and forgot to remove it. Now its removed from the code.